### PR TITLE
HIVE-28119: Iceberg: Allow insert clause with a column list in Merge query not_matched condition

### PR DESCRIPTION
**Pull request by deniskuzZ**
### What changes were proposed in this pull request?

Syntax improvement for the Merge query with insert clause

### Why are the changes needed?

Usability

### Does this PR introduce _any_ user-facing change?

No

### Is the change a dependency upgrade?

No

### How was this patch tested?

mvn test -Dtest=TestIcebergCliDriver -Dqfile=merge_iceberg_copy_on_write_partitioned.q,merge_iceberg_copy_on_write_unpartitioned.q

### CODE REVIEW

**Review by simhadri-g — APPROVED** [review]
LGTM +1

**Review by zhangbutao — APPROVED** [review]
LGTM +1

File: ql/src/java/org/apache/hadoop/hive/ql/parse/rewrite/CopyOnWriteMergeRewriter.java [hunk 1]
```diff
@@ -149,8 +153,20 @@ public void appendWhenNotMatchedInsertClause(MergeStatement.InsertClause insertC
         hintStr = null;
       }
       List<String> values = sqlGenerator.getDeleteValues(Context.Operation.MERGE);
-      values.add(insertClause.getValuesClause());
       
+      if (insertClause.getColumnListText() != null) {
+        String[] columnNames = insertClause.getColumnListText()
+            .substring(1, insertClause.getColumnListText().length() - 1).split(",");
+        String[] columnValues = insertClause.getValuesClause().split(",");
```
Comment (kasakrisz) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/rewrite/CopyOnWriteMergeRewriter.java:160]: Initially values are also represented in the AST
```
TOK_FUNCTION
  struct
  .
    TOK_TABLE_OR_COL
      some_alias
    a
  .
    TOK_TABLE_OR_COL
      some_alias
    c,c
```
We can try transform it to a `List<String>`

**Review by kasakrisz — CHANGES_REQUESTED** [review]

**Comment by deniskuzZ** [comment]
@kasakrisz, thanks for checking and sorry i didn't wait enough, I'll address your comments in the addendum

File: ql/src/java/org/apache/hadoop/hive/ql/parse/rewrite/CopyOnWriteMergeRewriter.java [hunk 2]
```diff
@@ -149,8 +153,20 @@ public void appendWhenNotMatchedInsertClause(MergeStatement.InsertClause insertC
         hintStr = null;
       }
       List<String> values = sqlGenerator.getDeleteValues(Context.Operation.MERGE);
-      values.add(insertClause.getValuesClause());
       
+      if (insertClause.getColumnListText() != null) {
+        String[] columnNames = insertClause.getColumnListText()
+            .substring(1, insertClause.getColumnListText().length() - 1).split(",");
```
[the remark this answers is not in the document]
Comment (deniskuzZ) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/rewrite/CopyOnWriteMergeRewriter.java:159]: 👍

**Pull request merged**